Temporary indexes

AudriusButkevicius opened the pull request:
Last commit is not finished, but throwing this out here to discuss the general approach before I spend more effort.
Also needs tests fixed and so on.

## TODO:
- [X] Send block index as an integer, rather than the whole block.
- [X] Add a flag to indicate this feature is off, don't send stuff if it's off.
- [X] Fallback to Request() in-case temporary request fails (no longer applicable, given we don't do validation, so it's just a if branch)
- [X] Address lint/vet/naming suggestions
- [X] Fix tests
- [X] Add tests

## Maybe:
- Limit number of entries we allow to receive, though people should just disable it.

*purpose*

> Files that are not fully downloaded are not shared, means that the initial seeding is always done from a single device which originates the change.
> Others start participating only when they have the full file.
>
> This changes that by advertising partially downloaded files (temporary).

AudriusButkevicius commented:
bump

AudriusButkevicius commented:
BTW, this is easier read commit by commit.

calmh commented:
Didn't look at the last commit yet. The rest looks surprisingly clean and noninvasive and matches something like what I'd expected. One thing I forgot to look for so I'm asking stupid questions instead - do we clear out the in progress information when a connection is closed and when we get an actual index update for a file (or there should always be a Forget update first?)?

AudriusButkevicius commented:
Builds fine, completely untested, needs adding tests and fixing tests.
Very close to the final thing, the last piece I need to add is fallbacking to a non-temporary Request() if the temporary one fails. 

Out of the two evils which one would you choose:
1. Storing old name as original name, and upon failure to open file, try again if name != originalName and flags == FromTemporary.
2. Calling Request within Request if FormTemporary flag is set (without it)

calmh commented:
Number two or three sound best to me.

One other thing; the temporary block list is kept in RAM right (on the side that receives the announcements)? This is fine, but should we have some sort of limit to how much data we keep there? Throwing it away doesn't affect functionality, only efficiency. I'm thinking the case where you're syncing a 100 GB file and send updated about that to a raspberry pi or something. Barring the variable block size, that might be a lot of blocks for the poor little thing to keep in RAM?

AudriusButkevicius commented:
Alternatively we could maintain the state in leveldb.

Also I should also add batched dispatch.

AudriusButkevicius commented:
Also a flag to ignore received updates.

AudriusButkevicius commented:
Only the following concerns left:
1. Ideally we should know that the other end has chosen to ignore temp indexes, and not waste bandwidth sending them.
2. As we've checked if the block is in the temp index, we could start dropping them? Or time based dropping?
3. Add tests.

calmh commented:
> Ideally we should know that the other end has chosen to ignore temp indexes, and not waste bandwidth sending them.

We could stash this in the clusterconfig options - maybe even as the inverse, i.e. the option is present if we want the temp updates? Not critical as far as I'm concerned.

> As we've checked if the block is in the temp index, we could start dropping them? Or time based dropping?

Maybe time based... Remove the oldest when we want to add something new and the number of blocks in the temp index stash reaches a limit?

Another thing... Ignore temp index updates for files where we already have the full new version?

AudriusButkevicius commented:
So time based doesn't really work because we now randomize pull order...
Regarding ignoring temp indexes for someone who has the latest file is a fair point, the only trouble is that you have to know that you have the latest file, which is racey.

AudriusButkevicius commented:
I guess I should add the relevant fields to ClusterConfig, as well as add `Name` field to it, to scrap the magic option we currently use for the Name field.

calmh commented:
Might as well, while you're at it, yeah.

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,99 @@
+	for _, update := range updates {
+		if update.UpdateType == protocol.UpdateTypeForget {
+			delete(p.files, update.Name)
+		} else if update.UpdateType == protocol.UpdateTypeAppend {
```
Could be a `switch update.UpdateType {` with a `default: panic("bug: unknown UpdateType")` possibly

calmh commented on lib/model/model.go:
```diff
@@ -730,11 +738,45 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
+}
+
+func (m *Model) requestFromTemporaryFile(deviceID protocol.DeviceID, folder, name string, offset int64, buf []byte) error {
+	name = defTempNamer.TempName(name)
```
We're skipping a lot of validation here. Can I accomplish any funky business by requesting a `../../../../something` with the temp flag, knowing what the TempName rewrite does?

AudriusButkevicius replied:
Good spot, but I guess we cannot really validate as the file might not yet be in our index to start with?

calmh replied:
We could do something to validate the resulting path. We had that in Request() earlier but I think it's been removed. I.e. something the checks that the full path after filepath.Clean() and whatnot actually has the folder path as a prefix.

calmh commented on lib/model/model.go:
```diff
@@ -995,6 +1049,21 @@ func (m *Model) IsPaused(device protocol.DeviceID) bool {
+	defer m.pmut.RUnlock()
+	result := make(map[string][]Connection, len(m.folderDevices))
+	for folder, devices := range m.folderDevices {
+		conns := make([]Connection, 0, len(devices))
```
May as well

```go
conns := make([]Connection, len(devices))
for i, device := range devices {
  conns[i] = m.conn[device]
}
```

calmh commented on lib/model/model.go:
```diff
@@ -1682,19 +1751,27 @@ func (m *Model) Availability(folder, file string) []protocol.DeviceID {
 	}
 
-	availableDevices := []protocol.DeviceID{}
+	infos := []AvailabilityInfo{}
```
`var infos []AvailabilityInfo`

calmh commented on lib/model/sentdownloadstate.go:
```diff
@@ -0,0 +1,145 @@
+// sentFolderDownloadState represents a state of what we've announced as available
+// to some remote device for a specific folder.
+type sentFolderDownloadState struct {
+	fileBlocks  map[string][]protocol.BlockInfo
```
This is kind of heavy. Could we not get away with just storing the highest block index we've sent or something like that?

calmh replied:
Maybe not, as the blocks may of course be out of order...

AudriusButkevicius replied:
Yeah, good idea.

calmh commented on lib/model/sharedpullerstate.go:
```diff
@@ -29,15 +30,18 @@ type sharedPullerState struct {
+	pullNeeded       int                  // Number of block pulls still pending
+	updated          time.Time            // Time when any of the counters above were last updated
+	closed           bool                 // True if the file has been finalClosed.
+	available        []protocol.BlockInfo // Blocks available in the temporary file
```
As above, maybe a list of offsets referring to the blocks already listed in the `file` attribute?

AudriusButkevicius replied:
Agree

calmh commented on lib/model/sentdownloadstate.go:
```diff
@@ -0,0 +1,145 @@
+// to some remote device for a specific folder.
+type sentFolderDownloadState struct {
+	fileBlocks  map[string][]protocol.BlockInfo
+	fileUpdated map[string]time.Time
```
Maps are kind of heavy and expensive to update as they may grow and need to copy themselves and so on. As such, it may be preferable to have a `map[string]someStruct` where the struct includes both the `time.Time` an the list-of-whatevers, instead of having separate maps for them.

calmh commented on lib/model/rwfolder.go:
```diff
@@ -1475,14 +1485,14 @@ func invalidateFolder(cfg *config.Configuration, folderID string, err error) {
-		if devices[i] == device {
-			devices[i] = devices[len(devices)-1]
-			return devices[:len(devices)-1]
+func removeDevice(infos []AvailabilityInfo, info AvailabilityInfo) []AvailabilityInfo {
```
No longer removes a specific device, but a specific AvailabilityInfo?

AudriusButkevicius commented:
Just a rebase.

AudriusButkevicius commented:
Added todo.

calmh commented:
Reviewed 1 of 2 files at r2, 3 of 5 files at r3, 3 of 9 files at r5.
Review status: 7 of 25 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[lib/model/devicedownloadstate.go, line 62 \[r5\]](https://reviewable.io:443/reviews/syncthing/syncthing/2252#-K2q8np2LkLFoZdWlnV_)** ([raw file](https://github.com/syncthing/syncthing/blob/43d65baa0f365a04ab51a2edbaf08cc1561d50b4/lib/model/devicedownloadstate.go#L62)):</sup>
Pretty sure this should be `map[string]*deviceFolderDownloadState` (i.e. map to pointer)
<!-- Reviewable comment -K2q8np2LkLFoZdWlnVa:-220779249 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/syncthing/syncthing/2252)
<!-- Sent from Reviewable.io -->

calmh commented:
Reviewed 7 of 10 files at r1, 3 of 9 files at r5.
Review status: 17 of 25 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[lib/model/sharedpullerstate.go, line 303 \[r5\]](https://reviewable.io:443/reviews/syncthing/syncthing/2252#-K2qDnmiMIztJvwE_hW0)** ([raw file](https://github.com/syncthing/syncthing/blob/43d65baa0f365a04ab51a2edbaf08cc1561d50b4/lib/model/sharedpullerstate.go#L303)):</sup>
"Available"
<!-- Reviewable comment -K2qDnmiMIztJvwE_hW1:-1440059882 -->

---

<sup>**[lib/protocol/message.go, line 127 \[r5\]](https://reviewable.io:443/reviews/syncthing/syncthing/2252#-K2qDZwqNxCdPGjxTogJ)** ([raw file](https://github.com/syncthing/syncthing/blob/43d65baa0f365a04ab51a2edbaf08cc1561d50b4/lib/protocol/message.go#L127)):</sup>
Needs a `// max:` of some kind
<!-- Reviewable comment -K2qDZwqNxCdPGjxTogK:1846946772 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/syncthing/syncthing/2252)
<!-- Sent from Reviewable.io -->

calmh commented:
So the status messages from reviewable are not really beautiful here, sorry about that... Just trying it out, and this is big enough where it makes sense.

AudriusButkevicius commented:
Can you by any chance move the unaddressed comments from the review.io business into github?

calmh commented:
I'll re-review with my previous comments in mind

calmh commented:
And feel free to try to merge master into this I guess, to resolve conflicts... I think a rebase would be rather painful here.

calmh commented on lib/model/model.go:
```diff
@@ -125,12 +130,14 @@ func NewModel(cfg *config.Wrapper, id protocol.DeviceID, deviceName, clientName,
 		folderStatRefs:    make(map[string]*stats.FolderStatisticsReference),
 		conn:              make(map[protocol.DeviceID]Connection),
-		deviceVer:         make(map[protocol.DeviceID]string),
+		config:            make(map[protocol.DeviceID]protocol.ClusterConfigMessage),
```
Naming; we have lots of things that are various kinds of config. `deviceClusterConf` or `deviceCC` or something maybe?

calmh commented on lib/model/model.go:
```diff
@@ -566,32 +573,32 @@ func (m *Model) folderSharedWith(folder string, deviceID protocol.DeviceID) bool
-		if addr != nil {
-			event["addr"] = addr.String()
+	// Only fire off event and print the line if it's the first ClusterConfig
+	// we are seeing.
```
General question; do we now send more than one clusterconfig per connection, or when would this happen?

AudriusButkevicius replied:
Never, planning for the future.

calmh commented on lib/model/model.go:
```diff
@@ -775,25 +781,49 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
+	// Cannot easily cache fd's because we might need to delete the file
+	// at any moment.
+
+	// Only check temp files if we the flag is set, and if we are set to
```
Typo; "we the flag"

calmh commented on lib/model/model.go:
```diff
@@ -775,25 +781,49 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
+			defer tempFile.Close()
+			_, err = tempFile.ReadAt(buf, offset)
+			if err != nil {
+				// The file might have finishe downloading, try the real file.
```
I think this part is a bit muddled, I was staring for too long at the err==nil/err!=nil logic, as it differs from the usual and couldn't just be processed automatically. :) How about breaking out functions `readFromTemporary`, `readFromCurrent` to read into buf or return an error or something to simplify this?

calmh commented:
You can ignore the rest of the old reviewable.io crap

calmh commented:
General thought; we have a bunch of places with `someVar&reallyLongFlagName == reallyLongFlagName` and similar; how about we standardize comparing flag-ANDs to zero instead when possible?

calmh commented on lib/model/model.go:
```diff
@@ -955,6 +1004,25 @@ func (m *Model) IsPaused(device protocol.DeviceID) bool {
 	return paused
 }
 
+func (m *Model) TemporaryIndexSubscribers() map[string][]Connection {
```
This thing worries me slightly, and the place in progressEmitter where it's called. It's a large double looped thing creating rather heavy structures and probably doing lots of allocations, called fairly frequently and then looped over. Is there some way we can precompute or simplify this so it's a little less work and allocations done on every announce interval?

AudriusButkevicius replied:
Good idea.

burkemw3 commented on lib/protocol/message.go:
```diff
@@ -148,6 +156,13 @@ type Device struct {
 	Options         []Option // max:64
 }
 
+type FileDownloadProgressUpdate struct {
```
The current BEP spec, mostly the Request message, appears to me to rely on folder name, file name, block offset, and block hash for identifying a block. What drove the desire to refer to blocks here with folder name, file name, file version, and block index number? It seems like a disconnect for me.

I'm not necessarily opposed to it but am currently curious. I also haven't been able to review this thoroughly enough for it to be immediately obvious right now.

AudriusButkevicius replied:
The fact that we are pulling the file implies that we have the have the index entry for it, so I guess we are just trying to make the messages as slim as possible.

I'd argue that the current implementation is not doing the right thing, and it should ideally refer to blocks by indexes/hashes.

burkemw3 replied:
Right, slim sounds like a good reason to me.

I do think it'd be good for the Request and FileDownloadProgressUpdate message types use consistent concepts, which might mean indexes instead of offsets. But, I'm not planning to badger you beyond this message.

calmh replied:
As a side note here, if (when) we do decide to change the wire format for requests, we must keep in mind that we will most likely be using variable blocks sizes soon (so index number may not be unambiguous) and probably different hash algorithms, which may not be guaranteed secure (so hash may not be unique). Hence folder+file+offset+size is the easiest way to specify what we need, and hash is included to verify it's the right thing plus give the opportunity to find another source for the same block *if* the hash algo is secure.

AudriusButkevicius commented:
Re: Ignore updates for files which we think are up to date (tricky, probably need to send version vector incase we get a temp index before a proper index from some other peer). The reason we want to know about all the blocks the device has is to potentially provide more granular inSyncBytes for remote devices.

This would be fairly expensive, as we'd have to perform potentially multiple index lookups to check what version of the file we've got.

calmh commented:
Do we need to ignore the updates though? It seems as if the cheapest would be to just accept the updates anyway, as they'll just live in RAM and probably be fairly short lived anyway

AudriusButkevicius commented:
So I think this is more or less complete.

There is a nil pointer panic here:
https://github.com/syncthing/syncthing/pull/2252/files#diff-526e0f13a297849163dbe36aec74c8c6R114

It seems to happen as you abruptly kill a connection, though I don't fully understand how a connection could become nil

AudriusButkevicius commented:
Looking forward to a round of reviews.

AudriusButkevicius commented:
BHUMP.

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+type deviceFolderFileDownloadState struct {
```
I'd love one or two sentences of comments on these types (this one and it's sibling below) on what they actually hold, as an intro the reader. :)

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+type deviceFolderFileDownloadState struct {
+	blockIndexes []int32
```
I think I would prefer this as `[]int` (simplest numeric type, doesn't have to be tied to the wire format as long as it's at least as big).

AudriusButkevicius replied:
Do we really want to go through the effort of type asserting?
This is an internal field so nobody should care about the type anyway.

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+
+type deviceFolderDownloadState struct {
+	mut   sync.RWMutex
+	files map[string]*deviceFolderFileDownloadState
```
I think this would be better as a `map[string]deviceFolderFileDownloadState` (that type sure is a mouthful). It makes it a little bit more annoying to update the map (need to assign the changed object back into the map instead of just accessing it via the pointer) but it saves maybe a whole bunch of pointers and makes the GC a bit happier.

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+					version:      update.Version,
+				}
+			} else if !local.version.Equal(update.Version) {
+				copy(local.blockIndexes, update.BlockIndexes)
```
Is there a reason to assume that `len(local.BlockIndexes) >= len(update.BlockIndexes)`? If not, the copy here might forget block indexes from the end.

AudriusButkevicius replied:
Good catch.

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+		}
+	}
+
+	p.mut.Unlock()
```
Might as well defer the unlock above instead

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+
+type deviceDownloadState struct {
+	mut     sync.RWMutex
+	folders map[string]*deviceFolderDownloadState
```
(note to self: my above preference for non-pointer values doesn't hold here due to the mutex inside deviceFolderDownloadState)

calmh commented on lib/model/devicedownloadstate.go:
```diff
@@ -0,0 +1,135 @@
+	p.mut.Unlock()
+}
+
+func (p *deviceFolderDownloadState) NumberOfBlocksInProgress() int {
```
I'm not sure how often this is intended to be called yet, but we could quite simply keep track of this in `Update()` instead and just expose this as a getter for `p.numBlocksInProgress` instead.

calmh commented on lib/model/model.go:
```diff
@@ -386,6 +392,8 @@ func (m *Model) Completion(device protocol.DeviceID, folder string) float64 {
 		return true
 	})
 
+	need -= int64(m.deviceDownloads[device].NumberOfBlocksInProgress() * protocol.BlockSize)
```
Don't you need to be holding `pmut` here?

calmh commented on lib/model/model.go:
```diff
@@ -606,31 +614,38 @@ func (m *Model) folderSharedWithUnlocked(folder string, deviceID protocol.Device
+	// Also, collect a list of folders we do share, and if he's interested in
+	// temporary indexes, subscribe the connection.
+
+	sharedFolders := make([]string, len(cm.Folders))
```
Should be `make([]string, 0, len(cm.Folders))`

calmh commented on lib/model/model.go:
```diff
@@ -657,12 +672,13 @@ nextFolder:
-
-	l.Infof(`Device %s client is "%s %s" named "%s"`, deviceID, cm.ClientName, cm.ClientVersion, cm.DeviceName)
+	if cm.Flags&protocol.FlagClusterConfigTemporaryIndexesDisabled == 0 {
+		m.progressEmitter.temporaryIndexSubscribe(conn, sharedFolders)
```
Comment that this breaks if we get more than one cluster config per connection, as we talk about that above

calmh commented on lib/protocol/message.go:
```diff
@@ -129,9 +129,17 @@ type ClusterConfigMessage struct {
 
+type DownloadProgressMessage struct {
+	Folder  string                       // max:64
+	Updates []FileDownloadProgressUpdate // max:1000000
```
We should decide whether this is all about DownloadProgress or FileDownloadProgress. I think DownloadProgress suffices.

AudriusButkevicius replied:
Not sure I understand.

calmh replied:
Just the wording difference between "DownloadProgressSomething" and "FileDownloadProgressSomething"; we could skip the "File" prefix.

AudriusButkevicius replied:
They are two different things though

calmh commented on lib/protocol/message.go:
```diff
@@ -159,6 +167,13 @@ type Device struct {
+	Name         string // max:8192
+	Version      Vector
+	BlockIndexes []int32 // max:1000000
+	UpdateType   uint32
```
I suggest moving this field to the top, as it kind of gives context to the whole thing and it feels sad to tuck it in the end. Not for any technical reason though.

calmh commented on lib/protocol/protocol.go:
```diff
@@ -51,14 +52,20 @@ const (
+// CLusterCOnfig flags
 const (
-	FlagRequestTemporary uint32 = 1 << iota
+	FlagClusterConfigTemporaryIndexesDisabled uint32 = 1 << 0
```
Why not make this just `FlagClusterConfigTemporaryIndexes`, with it being set meaning it is supported? That gives some conceptual compatibility with the current messages sent as they send this bit as a zero and don't support temporary indexes.

calmh commented on lib/model/model.go:
```diff
@@ -59,6 +59,11 @@ type service interface {
 	getState() (folderState, time.Time, error)
 }
 
+type AvailabilityInfo struct {
```
Could be just `type Availability`? (I forget why protocol.FileInfo and protocol.BlockInfo aren't called just protocol.File and protocol.Block but I *think* that's an unfortunate accident from like v0.2 where it was all in one package and a Block was something else already...)

calmh commented on lib/model/model.go:
```diff
@@ -1571,6 +1612,18 @@ func (m *Model) generateClusterConfig(device protocol.DeviceID) protocol.Cluster
 }
 
+func (m *Model) deviceVersion(id protocol.DeviceID) string {
+	cfg, ok := m.deviceClusterConf[id]
```
`// caller must hold pmut`

calmh commented:
Haven't yet looked at rwfolder / progressemitter, need to rest the brain a little. General question though: does this work on Windows, with the default file locking? Perhaps there is no default file locking? I'm thinking of the case when we try to open the temp file for reading while we already have it open for writing.

AudriusButkevicius commented:
The uint32 -> int hasn't been addressed, yet github says it's outdated due to the line moving due comments.

calmh commented:
rwfolder & progressemitter look sane to me too. We should not forget to run at least the ManyFiles and LargeFile2G benchmarks on this as well

AudriusButkevicius commented:
Not sure how to run that?
go run build.go bench doesn't run that.

AudriusButkevicius commented:
HEAD:
```

ab@benchrunner:~/gohome/src/github.com/syncthing/syncthing/test$ go test -v -tags "integration benchmark" -run 'Many'
=== RUN   TestBenchmarkTransferManyFiles
2016/02/04 16:44:13 Cleaning...
2016/02/04 16:44:13 Generating files...
2016/02/04 16:44:29 Total 154.5 MiB in 50000 files
2016/02/04 16:44:29 Starting instance 1...
2016/02/04 16:44:47 Starting instance 2...
2016/02/04 16:44:48 {8 2016-02-04 16:44:48.251666796 -0500 EST StateChanged map[from:idle to:scanning folder:default]}
2016/02/04 16:44:48 {11 2016-02-04 16:44:48.255608555 -0500 EST StateChanged map[duration:0.003941112 folder:default from:scanning to:idle]}
2016/02/04 16:44:48 {24 2016-02-04 16:44:48.758721406 -0500 EST StateChanged map[duration:0.503111959 folder:default from:idle to:syncing]}
2016/02/04 16:47:41 {100861 2016-02-04 16:47:41.730598454 -0500 EST StateChanged map[folder:default from:syncing to:idle duration:172.971874944]}
2016/02/04 16:47:41 Verifying...
2016/02/04 16:47:52 Result: Wall time: 2m52.971877048s
2016/02/04 16:47:52 Result: 0.9 MiB/s synced
2016/02/04 16:47:52 Receiver: Utime: 58.038599s
2016/02/04 16:47:52 Receiver: Stime: 12.148764s
2016/02/04 16:47:52 Receiver: MaxRSS: 83704 KiB
2016/02/04 16:47:52 Sender: Utime: 1m31.580588s
2016/02/04 16:47:52 Sender: Stime: 14.840326s
2016/02/04 16:47:52 Sender: MaxRSS: 94440 KiB
--- PASS: TestBenchmarkTransferManyFiles (219.35s)
PASS
ok      github.com/syncthing/syncthing/test     225.334s
ab@benchrunner:~/gohome/src/github.com/syncthing/syncthing/test$ go test -v -tags "integration benchmark" -run 'e2G'
=== RUN   TestBenchmarkTransferLargeFile2G
2016/02/04 16:17:58 Cleaning...
2016/02/04 16:17:58 Generating files...
2016/02/04 16:18:13 Total 2048.0 MiB in 1 files
2016/02/04 16:18:13 Starting instance 1...
2016/02/04 16:18:33 Starting instance 2...
2016/02/04 16:18:33 {8 2016-02-04 16:18:33.457956083 -0500 EST StateChanged map[folder:default from:idle to:scanning]}
2016/02/04 16:18:33 {11 2016-02-04 16:18:33.462342816 -0500 EST StateChanged map[duration:0.004387035 folder:default from:scanning to:idle]}
2016/02/04 16:18:33 {16 2016-02-04 16:18:33.689032277 -0500 EST StateChanged map[to:syncing duration:0.22668665600000001 folder:default from:idle]}
2016/02/04 16:20:09 {71 2016-02-04 16:20:09.44084549 -0500 EST StateChanged map[duration:95.751811987 folder:default from:syncing to:idle]}
2016/02/04 16:20:09 Verifying...
2016/02/04 16:20:17 Result: Wall time: 1m35.751813213s
2016/02/04 16:20:17 Result: 21.4 MiB/s synced
2016/02/04 16:20:17 Receiver: Utime: 50.466832s
2016/02/04 16:20:17 Receiver: Stime: 4.434617s
2016/02/04 16:20:17 Receiver: MaxRSS: 36424 KiB
2016/02/04 16:20:17 Sender: Utime: 50.195618s
2016/02/04 16:20:17 Sender: Stime: 6.685847s
2016/02/04 16:20:17 Sender: MaxRSS: 33348 KiB
--- PASS: TestBenchmarkTransferLargeFile2G (139.15s)
PASS
ok      github.com/syncthing/syncthing/test     139.163s
```

This branch:
```
=== RUN   TestBenchmarkTransferManyFiles
2016/02/04 16:39:35 Cleaning...
2016/02/04 16:39:35 Generating files...
2016/02/04 16:39:51 Total 154.5 MiB in 50000 files
2016/02/04 16:39:51 Starting instance 1...
2016/02/04 16:40:10 Starting instance 2...
2016/02/04 16:40:10 {8 2016-02-04 16:40:10.481336258 -0500 EST StateChanged map[folder:default from:idle to:scanning]}
2016/02/04 16:40:10 {11 2016-02-04 16:40:10.486709319 -0500 EST StateChanged map[duration:0.005373066 folder:default from:scanning to:idle]}
2016/02/04 16:40:11 {23 2016-02-04 16:40:11.01424029 -0500 EST StateChanged map[duration:0.527528878 folder:default from:idle to:syncing]}
2016/02/04 16:43:00 {100864 2016-02-04 16:43:00.370056931 -0500 EST StateChanged map[duration:169.355814943 folder:default from:syncing to:idle]}
2016/02/04 16:43:00 Verifying...
2016/02/04 16:43:11 Result: Wall time: 2m49.355816641s
2016/02/04 16:43:11 Result: 0.9 MiB/s synced
2016/02/04 16:43:11 Receiver: Utime: 57.128398s
2016/02/04 16:43:11 Receiver: Stime: 11.955964s
2016/02/04 16:43:11 Receiver: MaxRSS: 83576 KiB
2016/02/04 16:43:11 Sender: Utime: 1m30.825509s
2016/02/04 16:43:11 Sender: Stime: 14.270473s
2016/02/04 16:43:11 Sender: MaxRSS: 93544 KiB
--- PASS: TestBenchmarkTransferManyFiles (216.17s)
PASS
ok      github.com/syncthing/syncthing/test     218.871s
ab@benchrunner:~/gohome/src/github.com/syncthing/syncthing/test$ go test -v -tags "integration benchmark" -run 'e2G'
=== RUN   TestBenchmarkTransferLargeFile2G
2016/02/04 16:35:42 Cleaning...
2016/02/04 16:35:42 Generating files...
2016/02/04 16:35:55 Total 2048.0 MiB in 1 files
2016/02/04 16:35:55 Starting instance 1...
2016/02/04 16:36:14 Starting instance 2...
2016/02/04 16:36:15 {6 2016-02-04 16:36:15.334814861 -0500 EST StateChanged map[folder:default from:idle to:scanning]}
2016/02/04 16:36:15 {7 2016-02-04 16:36:15.337732919 -0500 EST StateChanged map[duration:0.002918852 folder:default from:scanning to:idle]}
2016/02/04 16:36:15 {16 2016-02-04 16:36:15.576951234 -0500 EST StateChanged map[to:syncing duration:0.239215493 folder:default from:idle]}
2016/02/04 16:37:51 {71 2016-02-04 16:37:51.609225597 -0500 EST StateChanged map[duration:96.032273031 folder:default from:syncing to:idle]}
2016/02/04 16:37:51 Verifying...
2016/02/04 16:37:59 Result: Wall time: 1m36.032274363s
2016/02/04 16:37:59 Result: 21.3 MiB/s synced
2016/02/04 16:37:59 Receiver: Utime: 49.207104s
2016/02/04 16:37:59 Receiver: Stime: 5.866334s
2016/02/04 16:37:59 Receiver: MaxRSS: 37644 KiB
2016/02/04 16:37:59 Sender: Utime: 48.915539s
2016/02/04 16:37:59 Sender: Stime: 7.552249s
2016/02/04 16:37:59 Sender: MaxRSS: 29948 KiB
--- PASS: TestBenchmarkTransferLargeFile2G (137.04s)
PASS
ok      github.com/syncthing/syncthing/test     137.058s

```

AudriusButkevicius commented:
So this does have 1 unaddressed comment (the integer harmonization) which we still need to discuss, and a bug, where the connection that registers for updates somehow becomes nil.

lkwg82 commented:
Just for my curiosity, can you give a little summary for what this is, PLZ. I dont get it.

AudriusButkevicius commented:
#950 

Files that are not fully downloaded are not shared, means that the initial seeding is always done from a single device which originates the change.
Others start participating only when they have the full file.

This changes that by advertising partially downloaded files (temporary).

AudriusButkevicius commented:
@calmh given your stuff has been deferred, should we try to get this in for 0.13?

AudriusButkevicius commented:
Addressed outstanding comments in Crucible, rebased on master.

calmh commented:
I've opened a new review CR-11, as the old one didn't survive the config changes I've done in crucible since. I've started looking at it and will do some benchmarking and testing as well later on.

calmh commented:
Got a panic with this revision while testing, at connection close (due to the other side shutting down):

```
[373HS] 09:44:08 INFO: Established secure connection to I6KAH76-66SLLLB-5PFXSOA-UFJCDZC-YAOMLEK-CP2GB32-BV5RQST-3PSROAU at 127.0.0.1:22003-127.0.0.1:54538 (direct-accept)
[373HS] 09:44:08 INFO: Device I6KAH76-66SLLLB-5PFXSOA-UFJCDZC-YAOMLEK-CP2GB32-BV5RQST-3PSROAU client is "syncthing v0.13.0-beta.2+28-ge5e8453-dirty-pr-2252" named "s1"
[373HS] 09:44:14 INFO: Established secure connection to MRIW7OK-NETT3M4-N6SBWME-N25O76W-YJKVXPH-FUMQJ3S-P57B74J-GBITBAC at 127.0.0.1:22003-127.0.0.1:54625 (direct-accept)
[373HS] 09:44:14 INFO: Device MRIW7OK-NETT3M4-N6SBWME-N25O76W-YJKVXPH-FUMQJ3S-P57B74J-GBITBAC client is "syncthing v0.13.0-beta.2+28-ge5e8453-dirty-pr-2252" named "s2"
[373HS] 09:48:41 INFO: Connection to I6KAH76-66SLLLB-5PFXSOA-UFJCDZC-YAOMLEK-CP2GB32-BV5RQST-3PSROAU closed: EOF
[373HS] 09:48:41 INFO: Connection to MRIW7OK-NETT3M4-N6SBWME-N25O76W-YJKVXPH-FUMQJ3S-P57B74J-GBITBAC closed: EOF
Panic at 2016-04-11T09:48:44+02:00
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x38 pc=0x2d3183]

goroutine 25 [running]:
panic(0x7540e0, 0xc820010130)
        /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/syncthing/syncthing/lib/model.(*ProgressEmitter).sendDownloadProgressMessages(0xc8202381e0)
        /Users/jb/src/github.com/syncthing/syncthing/lib/model/progressemitter.go:113 +0x273
github.com/syncthing/syncthing/lib/model.(*ProgressEmitter).Serve(0xc8202381e0)
        /Users/jb/src/github.com/syncthing/syncthing/lib/model/progressemitter.go:79 +0x761
created by github.com/syncthing/syncthing/lib/model.NewModel
        /Users/jb/src/github.com/syncthing/syncthing/lib/model/model.go:146 +0xbe8
```

calmh commented:
This seems to work nicely otherwise. I did a test with three devices, "s1" having a large file and being rate limited and "s2", "s3" pulling it. There's a fairly slow rampup of traffic between s2 and s3 though. I guess one thing that could be improved is that even though s3 might have a lot of blocks that s2 doens't, s2 just chugs along pulling blocks in random order and odds are it won't hit blocks that s3 has until s3 has a significant part of the file. I don't think we need to do it to merge this, but at some point it would probably be cool if we could periodically reshuffle the block list so that blocks with most peers are at the head of the list.

AudriusButkevicius commented:
Is pulling the block with most peers really the answer? That doesn't help distribution I guess, but I can't really gauge whats the best approach here.

calmh commented:
No, you're right, that's the wrong approach. Forget about that entirely for the moment. :)

alex2108 commented:
Could be pulling blocks with most peers and least peers at the same time be a solution? So If you pull 10 blocks at the same time (don't remember the default here) pull 5 blocks with the least peers and 5 blocks with most peers at the same time.

At least in theory this does not sound too hard to implement and should give a lot better performance.

Another thing that is a related problem: we have 3 device A,B and C all connected to each other.

-A adds a lot of files.
-B randomly starts pulling file1, C file2
-> since they are different files B and C do not exchange anything.
This means the same problem will happen when pulling the "wrong" files in random order so file pull order needs a similar optimization (in the example C also pulling file1 at the same time and B pulling file2)

calmh commented:
The last case is fine, because at some point they'll switch and then there are already two devices that have the files they need.

But I regret bringing up the block optimization thing. We'll deal with that in a future enhancement. No more discussion of it here, please. :)

AudriusButkevicius commented:
I believe I addressed all review comments, fixed the failing tests, as well as added a not so nice workaround for the crash.

AudriusButkevicius commented:
Anything left to be done?

rumpelsepp commented:
It seems that AppVeyor has failed.

AudriusButkevicius commented:
I can't recall seeing it pass.

calmh commented:
I need to retest it a little. Other than that, should be fine. There are apparently conflicts blocking appveyor, although our stuff seems to think that's manageable...

calmh commented:
I'm getting some test failures in the integration tests that i don't think we had before, going to investigate that a little.

AudriusButkevicius commented:
I though that was a dead horse, but I'll try running them.

calmh commented:
When restarting a device, another device crashes with this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x38 pc=0x2d32c3]

goroutine 25 [running]:
panic(0x754260, 0xc820010120)
        /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/syncthing/syncthing/lib/model.(*ProgressEmitter).sendDownloadProgressMessages(0xc82011e240)
        /Users/jb/src/github.com/syncthing/syncthing/lib/model/progressemitter.go:114 +0x273
github.com/syncthing/syncthing/lib/model.(*ProgressEmitter).Serve(0xc82011e240)
        /Users/jb/src/github.com/syncthing/syncthing/lib/model/progressemitter.go:80 +0x761
created by github.com/syncthing/syncthing/lib/model.NewModel
        /Users/jb/src/github.com/syncthing/syncthing/lib/model/model.go:146 +0xbec
```

calmh commented:
They are a dead horse, generally speaking, but I still use the SyncCluster*, Benchmark*, Restart* and some others as quick smoke tests.

AudriusButkevicius commented:
Ok, now really fixed the crash.

AudriusButkevicius commented:
Rebased

calmh commented:
I can't get this to crash now so that's better. :) I'm going to hammer a little bit more and see if I can actually understand some of the parts of the code that I currently dont, then I think we can merge this today \o/

calmh commented:
In the meantime, please do feel free to write up the relevant changes on the spec ;)

AudriusButkevicius commented:
Yeah, put that on my plate in the forum post, as well as the hello messgage stuff, and I'll write it up.

calmh commented:
@st-review merge this thing, will you please?

lib/model, lib/protocol: Implement temporary indexes (fixes #950)

st-review commented:
:ok_hand: Merged as 1a5f524ae4a47588ed633eff78097c6255aa2bcf. Thanks, @AudriusButkevicius!

canton7 commented:
:+1: :smile:

AudriusButkevicius commented:
Only took a year.

AudriusButkevicius commented:
https://github.com/syncthing/syncthing/pull/1271

A year and 3 months that is.

calmh commented:
:)

calmh commented:
I don't think this takes care of #2842 currently right, as we don't use the temp indexes received to bump the completed percentage for a remote device?

(I'm not 100% sure we *should* either, but that's a discussion for that issue itself.)

AudriusButkevicius commented:
https://github.com/syncthing/syncthing/pull/2252/commits/440d4552e3e1b682e19120d1f1f24a01ce71c2b5

AudriusButkevicius commented:
I think it might jump around though, between index update and marking temp file as done.

calmh commented:
Ah, right. Close that one too then.
